pkg/reloader/vault.go

```diff
@@ -0,0 +1,173 @@
+
+	vaultConfig.Path = os.Getenv("VAULT_PATH")
+	if vaultConfig.Path == "" {
+		vaultConfig.Path = "kubernetes"
+	}
+
+	vaultConfig.Namespace = os.Getenv("VAULT_NAMESPACE")
+	if vaultConfig.Namespace == "" {
+		vaultConfig.Namespace = "default"
+	}
+
+	vaultConfig.SkipVerify, _ = strconv.ParseBool(os.Getenv("VAULT_SKIP_VERIFY"))
+
+	vaultConfig.TLSSecret = os.Getenv("VAULT_TLS_SECRET")
+
+	vaultConfig.TLSSecretNS = os.Getenv("VAULT_TLS_SECRET_NS")
+	if vaultConfig.TLSSecretNS == "" {
+		vaultConfig.TLSSecretNS = "default"
+	}
+
+	vaultConfig.ClientTimeout, _ = time.ParseDuration(os.Getenv("VAULT_CLIENT_TIMEOUT"))
+	if vaultConfig.ClientTimeout == 0 {
+		vaultConfig.ClientTimeout = 10 * time.Second
+	}
+
+	return &vaultConfig
+}
+
+func (c *Controller) initVaultClient(config *VaultConfig) error {
+	c.logger.Info("Creating Vault client")
+
+	clientConfig := vaultapi.DefaultConfig()
+	if clientConfig.Error != nil {
+		return clientConfig.Error
+	}
+
+	clientConfig.Address = config.Addr
+	clientConfig.Timeout = config.ClientTimeout
+
+	tlsConfig := vaultapi.TLSConfig{Insecure: config.SkipVerify}
+	err := clientConfig.ConfigureTLS(&tlsConfig)
+	if err != nil {
+		return err
+	}
+
+	if config.TLSSecret != "" {
+		tlsSecret, err := c.kubeClient.CoreV1().Secrets(config.TLSSecretNS).Get(
+			context.Background(),
+			config.TLSSecret,
+			metav1.GetOptions{},
+		)
+		if err != nil {
+			return fmt.Errorf("failed to read Vault TLS Secret: %s", err)
+		}
+
+		clientTLSConfig := clientConfig.HttpClient.Transport.(*http.Transport).TLSClientConfig
+
+		pool := x509.NewCertPool()
+
+		ok := pool.AppendCertsFromPEM(tlsSecret.Data["ca.crt"])
+		if !ok {
+			return fmt.Errorf("error loading Vault CA PEM from TLS Secret: %s", tlsSecret.Name)
+		}
+
+		clientTLSConfig.RootCAs = pool
+	}
+
+	vaultClient, err := vault.NewClientFromConfig(
+		clientConfig,
+		vault.ClientRole(config.Role),
+		vault.ClientAuthPath(config.Path),
+		vault.ClientAuthMethod(config.AuthMethod),
+		vault.ClientLogger(logrusadapter.NewFromEntry(c.logger)),
+		vault.VaultNamespace(config.Namespace),
+	)
+	if err != nil {
+		return err
+	}
+
+	// Check connection to Vault
+	health, err := vaultClient.RawClient().Sys().Health()
+	c.logger.Debugf("Vault health: %v", health)
+	if err != nil {
+		c.logger.Error("testing connection to Vault failed")
+		return err
+	}
+
+	c.vaultClient = vaultClient.RawClient()
+	c.logger.Info("Vault client initialized")
+	return nil
+}
+
+type ErrSecretNotFound struct {
+	secretPath string
+}
+
+func (e ErrSecretNotFound) Error() string {
+	return fmt.Sprintf("Vault secret %s not found", e.secretPath)
+}
+
+func getSecretVersionFromVault(vaultClient *vaultapi.Client, secretPath string) (int, error) {
+	secret, err := vaultClient.Logical().Read(secretPath)
+	if err != nil {
+		return 0, err
+	}
+	if secret != nil {
+		secretVersion, err := secret.Data["metadata"].(map[string]interface{})["version"].(json.Number).Int64()
+		if err != nil {
+			return 0, err
```

Comment:
Should this be a terminating error because the reloader will exit if this returns

Reply:
I modified the error handling part in the method that calls this one because at that point we should have a working connection with Vault so we don't need to terminate the process just log the error and move on to the next secret path...